subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c

```diff
@@ -375,39 +412,87 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
 	const struct flash_area *fa;
 	int area_id = img_mgmt_flash_area_id(slot);
 	int rc = 0;
-	bool active = (slot == img_mgmt_active_slot(img_mgmt_active_image()));
-
-#if defined(CONFIG_MCUMGR_GRP_IMG_STATUS_HOOKS)
-	int32_t err_rc;
-	uint16_t err_group;
+	/* image the requested slot is defined within */
+	int image = img_mgmt_slot_to_image(slot);
+	/* active_slot is slot that is considered active/primary/executing
+	 * for a given image.
+	 */
+	int active_slot = img_mgmt_active_slot(image);
+	enum img_mgmt_next_boot_type type;
+	int next_boot_slot = img_mgmt_get_next_boot_slot(image, &type);
+
+	LOG_DBG("(%d, %s)", slot, confirm ? "confirm" : "test");
+	LOG_DBG("aimg = %d, img = %d, aslot = %d, slot = %d, nbs = %d",
+		img_mgmt_active_image(), image, active_slot, slot, next_boot_slot);
+
+	/* MCUmgr should not allow to confirm non-active image slots to prevent
+	 * confirming something that might not have been verified to actually be bootable
+	 * or have stuck in primary slot of other image. Unfortunately there was
+	 * a bug in logic that always allowed to confirm secondary slot of any
+	 * image. Now the behaviour is controlled via Kconfig options.
+	 */
+#ifndef CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY
+	if (confirm && image != img_mgmt_active_image() &&
+	    (!IS_ENABLED(CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY) ||
+	     slot == active_slot)) {
+		LOG_DBG("Not allowed to confirm non-active images");
+		return IMG_MGMT_ERR_INVALID_SLOT;
+	}
 #endif
 
-	if (active) {
-		confirm = true;
-	} else {
-		if (slot != 1 &&
-		    !(CONFIG_MCUMGR_GRP_IMG_UPDATABLE_IMAGE_NUMBER == 2 && slot == 3)) {
-			return MGMT_ERR_EINVAL;
-		}
+	/* Setting test to active slot is not allowed. */
+	if (!confirm && slot == active_slot) {
+		return IMG_MGMT_ERR_INVALID_SLOT;
```

Comment:
I think we should use proper errors for these, the slot isn't invalid, the current state disallows this

Reply:
Defined new error code.

Reply:
Still seems to be using invalid slot error

Reply:
Sorry, fixed wrong line. Now both lines are fixed and two new error codes added.